schema/python/schema/codegen.py

```diff
@@ -0,0 +1,318 @@
+import re
+import subprocess
+from collections import namedtuple
+from .ir import *
+from .registry import lookup
+from . import typing as ty
+import json
+
+class ExprFunctor:
```

Comment:
This might be a stylistic thing, but I don't think we need to write a visitor for this. We could just have 2 or 3 recursive functions instead. I think the benefit of using a couple functions instead of a visitor is that it is much clearer what is going on.

Reply:
given there are a few subclasses below i think it is fine in this case to reuse the dispatching logic

Reply:
those logic is useful to capture the base class's fields

Reply:
It just seems that every place you call visit you know exactly what you are visiting (because there is no subclassing). For example, in
```
for field in obj.fields:
    fields.append(fields_comment[field.name])
    fields.append(self.visit(field))
```
we could just replace `self.visit(field)` with `self.visit_field_def(field)`. I believe the same is true of every other place `self.visit` is used.